server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,310 @@
+    }
+
+    protected String executorName() {
+        return ThreadPool.Names.GENERIC;
+    }
+
+    private void setMaxConcurrentFetches(Integer maxConcurrentFetches) {
+        this.maxConcurrentFetches = maxConcurrentFetches;
+    }
+
+    @Override
+    public SnapshotShardSizeInfo snapshotShardSizes() {
+        return new SnapshotShardSizeInfo(knownSnapshotShardSizes);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
+            final Set<SnapshotShard> onGoingSnapshotRecoveries = new HashSet<>();
+
+            int unknownShards = 0;
+            for (ShardRouting shardRouting : event.state().routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+                if (shardRouting.primary() && shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+                    final RecoverySource.SnapshotRecoverySource snapshotRecoverySource =
+                        (RecoverySource.SnapshotRecoverySource) shardRouting.recoverySource();
+                    final SnapshotShard snapshotShard = new SnapshotShard(snapshotRecoverySource.snapshot(),
+                        snapshotRecoverySource.index(), shardRouting.shardId());
+                    onGoingSnapshotRecoveries.add(snapshotShard);
+
+                    // check if already populated entry
+                    if (knownSnapshotShardSizes.containsKey(snapshotShard) == false) {
+                        // check if already fetching snapshot info in progress
+                        if (unknownSnapshotShards.add(snapshotShard)) {
+                            queue.add(snapshotShard);
+                            unknownShards += 1;
+                        }
+                    }
+                }
+            }
+
+            // Clean up keys from knownSnapshotShardSizes that are no longer needed for recoveries
+            synchronized (this) {
+                isMaster = true;
+                ImmutableOpenMap.Builder<SnapshotShard, Long> newSnapshotShardSizes = null;
+                for (ObjectCursor<SnapshotShard> shard : knownSnapshotShardSizes.keys()) {
+                    if (onGoingSnapshotRecoveries.contains(shard.value) == false) {
+                        if (newSnapshotShardSizes == null) {
+                            newSnapshotShardSizes = ImmutableOpenMap.builder(knownSnapshotShardSizes);
+                        }
+                        newSnapshotShardSizes.remove(shard.value);
+                    }
+                }
+                if (newSnapshotShardSizes != null) {
+                    knownSnapshotShardSizes = newSnapshotShardSizes.build();
+                }
+            }
+
+            final int nbFetchers = Math.min(unknownShards, maxConcurrentFetches);
+            for (int i = 0; i < nbFetchers; i++) {
+                final int activeFetches = concurrentFetches.incrementAndGet();
+                if (activeFetches < maxConcurrentFetches + 1) {
```

Comment:
I think there is a race here, in that if all threads polled and found out that `queue` is empty but did not count down the counter yet, we could add to the queue here and risk that the increment results in a number higher than `maxConcurrentFetches`?

Reply:
Right sorry for missing that ... looking at this again, do we even have to be so tricky here? :) Maybe we should just sychronize `fetchNextSnapshotShard` and this block:

```java
            for (int i = 0; i < nbFetchers; i++) {
                final int activeFetches = concurrentFetches.incrementAndGet();
                if (activeFetches < maxConcurrentFetches + 1) {
                    fetchNextSnapshotShard();
                } else {
                    final int value = concurrentFetches.decrementAndGet();
                    assert value >= 0 : "Unexpected value: " + value;
                }
            }
```

on the same mutex (e.g.  on `queue`) and use a normal `int` for the in-progress count (no need to bother with the count-down here then as well, we can just call `fetchNextSnapshotShard()` however often we are allowed in a loop and just adjust the in-progress count once)?  That seems the simplest and least error prone since the performance of the locking is not relevant here anyway?

Reply:
Following Armin's opinion for simplicity and usage of synchronized blocks I reworked the concurrency part in 674d6c9387d90f492eaf3d04bd573616b06d5327. I hope it makes things simpler but I must admit that I don't have clear ideas today so I'd appreciate more reviews there.